src/main/java/uk/gov/ch/oauth/IdentityProvider.java

```diff
@@ -0,0 +1,84 @@
+package uk.gov.ch.oauth;
+
+import java.util.Base64;
+import java.util.Base64.Decoder;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import uk.gov.companieshouse.environment.EnvironmentReader;
+import uk.gov.companieshouse.logging.Logger;
+import uk.gov.companieshouse.logging.LoggerFactory;
+
+@Component
+public class IdentityProvider implements IIdentityProvider {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger("docs.developer.ch.gov.uk");
+    private final static Decoder decoder = Base64.getDecoder();
+    final private byte[] requestKey;
+    final private String authorizationUri;
+    final private String clientId;
+    final private String redirectUri;
+    final private String clientSecret;
+
+    @Autowired
+    public IdentityProvider(final EnvironmentReader reader) {
+        requestKey = decoder.decode(reader.getMandatoryString("DEVELOPER_OAUTH2_REQUEST_KEY"));
+        this.authorizationUri = reader.getMandatoryString("OAUTH2_AUTH_URI");
+        this.clientId = reader.getMandatoryString("CHS_DEVELOPER_CLIENT_ID");
+        redirectUri = reader.getMandatoryString("OAUTH2_REDIRECT_URI");
+        clientSecret = reader.getMandatoryString("CHS_DEVELOPER_CLIENT_SECRET");
+       // TODO is this better? redirectUri = reader.getMandatoryString("REDIRECT_URI");
```

Comment:
Just to add to the first comment, I've left a comment under where you've done the redirect and why I think it's probably not the correct way to do it. So it might be worth keeping the the current version for the time being.

Reply:
Can we uncomment this line out please, as it will be used when correcting the redirect.

Reply:
I wasn't finding the value in my Vagrant environment, but was finding the OAUTH... value, which I suspect should be the correct one, however I noticed @Zain-Abbas1  using the other key in your changes.

Reply:
The value should be in your vagrant env, you may need to do a git pull. `OAUTH2_REDIRECT_URI` is the callback redirect, `REDIRECT_URI` is where the app should redirect to once the callback is complete. This is what I mentioned before about me needing to do it this way until the ticket is picked up from the backlog, that will investigate the redirect in a proper way.

Reply:
At present, the code works out the redirect for return using the source of the original signing request.

I'm going to have to leave the decision in your hands.